src/data-analysis-environments.rst

```diff
@@ -0,0 +1,322 @@
+By doing so, it embeds a lightweight data science environment directly into the data with live previews and visualizations of the computing results.
+To lower the barriers of learning, DS.js further generates API suggestions to manipulate the selected data piece in the table.
+In addition, DS.js encapsulates all the user-generated code into a single URL for sharing.
+The first-use evaluation with 8 subjects proofs the usability of the system.
+
+.. image:: https://bearzx.github.io/DS.js/imgs/open-dsjs.gif
+  :width: 800
+  :alt: Loading DS.js to any HTML tables
+
+.. container:: bib-item
+
+  .. bibliography:: data-analysis-environments.bib
+    :filter: key == 'zhang:2017'
+
+This idea has been further explored in **Mallard** :cite:`zhang:2019`, which investigated how to turn web as a contextualized prototyping environment for machine learning.
+Mallard is shown to be useful for hobbyist-level machine learning prototyping through a set of case studies, including augmenting social media sites with sentiment analysis, performing style transfer on Google Image search.
+
+.. container:: bib-item
+
+  .. bibliography:: data-analysis-environments.bib
+    :filter: key == 'zhang:2019'
+
+Tools for Cleaning and Integrating
+==================================
+Often times, data science workers receive poor-structured and incomplete datasets.
+The datasets must be cleaned or redesigned to meet the requirement of the algorithms or models.
+It is said that data scientists spend most of the time collecting and cleaning the data rather than mining and modeling data.
+The challenge here is that there might be variants of situations that need to be considered. 
+For example, a missing value may have different representations (N/A, null, '', undefined).
+Data science workers may develop their own strategies to handle missing data (N/As) - deleting a record if it has too many missing attributes; using min/max/mean/median/mod to fill numerical missing attributes.
+Data science workers may need to normalize numerical data, remove non-related attributes and outlier records, and encode discrete attributes to numerical values.
+It is a tedious process to iterate over all possibilities.
+It is also difficult to specify the transformation and reuse it across different contexts.
+
+Interactive Visual Specification
+--------------------------------
+What if the system can prompt users for these transformations without asking them to write code?
+Wrangler used rule-based inference to interactively suggest users valid transformations based on their current selection :cite:`kandel:2011`.
+All the transformations are represented in natural language to reduce the cost of writing and reading code.
+Users can specify parameters in the transformations through direct editing.
+In addition, Wrangler captures the editing histories of the transformation.
+In this way, data science workers can improve their efficiency comparing to manual editing.
```

Comment:
If the paper provided evidence that this improves worker productivity, summarize that evidence here. If not, say it is a conjecture without evidence.